Fix updater workflow

joshuagl opened the pull request:
**Fixes issue #**: N/A

**Description of the changes being introduced by the pull request**:
The updater was not adhering to the detailed client workflow. Specifically newly downloaded root metadata was not being verified with a threshold of signatures from itself.

Per the detailed client workflow in the specification step 1.2
> Version N+1 of the root metadata file MUST have been signed by:
(1) a threshold of keys specified in the trusted root metadata file (version N), and
(2) a threshold of keys specified in the new root metadata file being validated (version N+1)."

This PR ensures that point two is verified by the updater.

**Please verify and check that the pull request fulfills the following
requirements**:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature

sechkova committed:
Add updater tests for root signed with current keys

Co-authored-by: Joshua Lock <jlock@vmware.com>
Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>
Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl committed:
updater: rename _verify_uncompressed_metadata_file

We no longer handle compressed metadata files, so rename this method to be
shorter and less confusing:
_verify_uncompressed_metadata_file -> _verify_metadata_file

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl committed:
updater: remove redundant __verify_root_chain_link method

This method is duplicating verification steps which have already been
completed before the file was written to disk.

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl committed:
updater: verify newly downloaded root metadata with its signatures

Per the detailed client workflow in the specification step 1.2

"Version N+1 of the root metadata file MUST have been signed by:
(1) a threshold of keys specified in the trusted root metadata file
(version N), and
(2) a threshold of keys specified in the new root metadata file being
validated (version N+1)."

Number 2 is implemented here as this step was not being performed by the
Updater. Unfortunately we can't use existing signature verification
methods in tuf.sig, because tuf.sig.signature_status() does not verify
signatures for keys which are not listed in keydb (and tuf.sig.verify
uses tuf.sig.signature_status)

Therefore this patch introduces a method for verifying signatures with
root keys listed in the signable being verified.

Signed-off-by: Joshua Lock <jlock@vmware.com>

trishankatdatadog commented on tuf/client/updater.py:
```diff
@@ -1379,12 +1379,50 @@ def verify_target_file(target_file_object):
+      # The ANYKEY_SCHEMA check in verify_signature expects the keydict to
+      # include a keyid
+      key['keyid'] = keyid
+      valid_sig = securesystemslib.keys.verify_signature(key, signature, signed)
```
Should we bail on a single exception here? Feels like we need a try-catch...

joshuagl replied:
`_get_metadata_file` has some logic to collect all exceptions and pack them into `NoWorkingMirrorError`, so I *think* this just continues the existing pattern?

trishankatdatadog commented on tuf/client/updater.py:
```diff
@@ -1379,12 +1379,50 @@ def verify_target_file(target_file_object):
 
 
-  def _verify_uncompressed_metadata_file(self, metadata_file_object,
+  def _verify_root_sigs(self, signable):
```
I think we need a better name, like `_self_verify_root_sigs()`, because it wasn't clear where the keys are coming from (the new root metadata file itself, which seems dangerous until you consider that this should be called only _after_ verifying that new follows from old, then you check that new follows from new).

jku replied:
`_verify_root_self_signed()`?

joshuagl replied:
good suggestion, I like it

trishankatdatadog commented on tuf/client/updater.py:
```diff
@@ -1448,6 +1486,14 @@ def _verify_uncompressed_metadata_file(self, metadata_file_object,
+    # new threshold of new signatures contained within the downloaded root
+    # metadata object
+    if valid and metadata_role == 'root':
+      valid = self._verify_root_sigs(metadata_signable)
```
I also wouldn't call this function inside _verify_metadata_file(), because there is too much magic. Best to call it explicitly in the _update_root_metadata() code IMHO.

joshuagl replied:
Agreed, this does make `_verify_metadata_file()` much more magical. I opted to call it inside `_verify_metadata_file()` because this enable us to perform the check before the tempfile is persisted, and makes it easier to perform the check per download mirror.

If we do the check in `_update_root_metadata()` we need to clean up a file from the clients metadata directory and, perhaps less importantly, can only perform the check on a single file returned by `_update_root_metadata()`

jku replied:
+1. I think the point of never writing unverified files is _critical_ and should probably be encoded in comments or docs... If we do not adhere to that it's almost impossible to ensure that our state of metadata stays valid.

trishankatdatadog replied:
Also, we want to make sure that other functions that depend on this function do the same thing. Let's leave the magic alone for now, but document clearly why it is there, and not somewhere else.

trishankatdatadog commented on tuf/client/updater.py:
```diff
@@ -1811,9 +1839,6 @@ def _update_metadata(self, metadata_role, upperbound_filelength, version=None):
     updated_metadata_object = metadata_signable['signed']
     current_metadata_object = self.metadata['current'].get(metadata_role)
 
-    self._verify_root_chain_link(metadata_role, current_metadata_object,
```
Why did we remove this code again?

joshuagl replied:
It's just duplicating steps that have already been performed in `_verify_metadata_file` and not performing those steps until after the tempfile has been persisted.

trishankatdatadog commented on tests/test_updater_root_rotation_integration.py:
```diff
@@ -219,6 +219,80 @@ def test_root_rotation(self):
 
 
 
+  def test_BUG_ROOT_NOT_VERIFIED_WITH_CURRENT_KEYIDS_AND_THRESHOLD(self):
```
Hold on: according to the documentation, this test should fail. Why doesn't it?

joshuagl replied:
Good catch, the doc string needs updating now that the fix is implemented.

trishankatdatadog commented on tests/test_updater_root_rotation_integration.py:
```diff
@@ -413,8 +495,112 @@ def test_root_rotation_unmet_threshold(self):
+
+
+
+  def test_root_rotation_discard_untrusted_version(self):
```
Lovely test.

trishankatdatadog commented on tests/test_updater_root_rotation_integration.py:
```diff
@@ -413,8 +495,112 @@ def test_root_rotation_unmet_threshold(self):
+
+
+
+  def test_root_rotation_unmet_new_threshold(self):
```
Lovely test, but there's some overlap between this test and the next. Maybe clarify that there is at least one successful intermediate rotation involved or something.

trishankatdatadog commented on tests/test_updater_root_rotation_integration.py:
```diff
@@ -219,6 +219,80 @@ def test_root_rotation(self):
+
+
+
   def test_root_rotation_full(self):
```
Hold on: how does this test full rotation? We're still at version 1 of root at the end.

jku replied:
Are you mixing up tests? `test_root_rotation_full()` seems fine to me and was not touched in this commit.

jku commented on tests/test_updater_root_rotation_integration.py:
```diff
@@ -342,22 +415,30 @@ def test_root_rotation_missing_keys(self):
+      url_prefix = self.repository_mirrors['mirror1']['url_prefix']
+      url_file = os.path.join(url_prefix, 'metadata', '2.root.json')
 
-        # Verify that '2.root.json' is the culprit.
-        self.assertEqual(url_file.replace('\\', '/'), mirror_url)
-        self.assertTrue(isinstance(mirror_error,
+      # Verify that '2.root.json' is the culprit.
+      self.assertEqual(url_file.replace('\\', '/'), mirror_url)
```
I know this is how most tests do this but replacing the 3 lines with this should do the job and makes the test more readable:
`self.assertTrue(mirror_url.endswith("/2.root.json"))`

(same comment applies to other modified/added tests)

joshuagl replied:
Thanks for this suggestion, I've pushed a patch implementing it: adbb5fe

jku commented on tests/test_updater_root_rotation_integration.py:
```diff
@@ -413,8 +494,112 @@ def test_root_rotation_unmet_threshold(self):
+      self.repository_updater.refresh()
+
+    for mirror_url, mirror_error in six.iteritems(cm.exception.mirror_errors):
+      url_prefix = self.repository_mirrors['mirror1']['url_prefix']
+      url_file = os.path.join(url_prefix, 'metadata', '3.root.json')
+
+      # Verify that '2.root.json' is the culprit.
+      self.assertEqual(url_file.replace('\\', '/'), mirror_url)
```
same comment as above

jku commented on tests/test_updater_root_rotation_integration.py:
```diff
@@ -413,8 +481,103 @@ def test_root_rotation_unmet_threshold(self):
+      self.repository_updater.refresh()
+
+    for mirror_url, mirror_error in six.iteritems(cm.exception.mirror_errors):
+      # Verify that '2.root.json' is the culprit.
+      self.assertTrue(mirror_url.endswith('/3.root.json'))
```
Now that the code is simpler it's easy to see an issue with the comment :)
Could even drop the comment: code seems fairly obvious to me...

joshuagl replied:
🤦 thank you for the diligent review. Fixed in 6eaf194

jku commented on tests/test_updater_root_rotation_integration.py:
```diff
@@ -413,8 +481,103 @@ def test_root_rotation_unmet_threshold(self):
+      self.repository_updater.refresh()
+
+    for mirror_url, mirror_error in six.iteritems(cm.exception.mirror_errors):
+      # Verify that '2.root.json' is the culprit.
+      self.assertTrue(mirror_url.endswith('/3.root.json'))
```
comment clashes with code here too

joshuagl commented:
I believe I've addressed all review comments. Once the PR is approved I'd like to squash some of the changes addressing review comments into the patches that introduced the changes.

trishankatdatadog reviewed (approved):
LGTM
